layers/igmp.go

```diff
@@ -16,18 +16,210 @@ import (
 
 type IGMPType uint8
 
+const (
+	MembershipQuery    IGMPType = 0x11 // General or group specific query
+	MembershipReportV1 IGMPType = 0x12 // Version 1 Membership Report
+	MembershipReportV2 IGMPType = 0x16 // Version 2 Membership Report
+	LeaveGroup         IGMPType = 0x17 // Leave Group
+	MembershipReportV3 IGMPType = 0x22 // Version 3 Membership Report
+)
+
+// String conversions for IGMP message types
+func (i IGMPType) String() string {
+	switch i {
+	case MembershipQuery:
+		return "IGMP Membership Query"
+	case MembershipReportV1:
+		return "IGMPv1 Membership Report"
+	case MembershipReportV2:
+		return "IGMPv2 Membership Report"
+	case MembershipReportV3:
+		return "IGMPv3 Membership Report"
+	case LeaveGroup:
+		return "Leave Group"
+	default:
+		return ""
+	}
+}
+
+type IGMPQuery interface {
+	decodeResponse(data []byte) error
+}
+
 // IGMP is the packet structure for IGMP messages.
 type IGMP struct {
 	BaseLayer
-	Type            IGMPType
```

Comment:
> This change to the IGMP type breaks backwards compatibility.....  We try very hard to not break existing code with new commits.

got it and makes sense. Great stance to take.

> Is the current IGMP struct version-specific

It is not. But the current IGMP implementation only calls out a single IGMP query type. That being When [byte 0 == 0x11](https://github.com/google/gopacket/blob/master/layers/igmp.go#L53 ) ( a Membership query) . This type happens to be shared between all 3 versions of IGMP. Version inference of the membership query type is then made based on content length of the packet and contents of byte 1.

I am guessing that IGMP v1 is rarely used so it probably safe to ignore, but V2 still has some use. I will revisit this pull request and see if it can be squeezed into the current implementation.